plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,289 @@
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  protected getKubernetesRequestedClusters(
+    req: Request,
+  ): KubernetesProxyClusters {
+    const encodedClusters: string =
+      req.header(HEADER_KUBERNETES_CLUSTERS) ?? '';
+
+    if (!encodedClusters) {
+      return {};
+    }
+
+    try {
+      const decodedClusters = Buffer.from(encodedClusters, 'base64').toString();
+      const clusters: KubernetesProxyClusters = JSON.parse(decodedClusters);
+      return clusters;
+    } catch (e: any) {
+      this.logger.error(
+        `error with encoded cluster header: ${JSON.stringify(e)}`,
+      );
+    }
+    return {};
+  }
+
+  protected async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const authMetadata: KubernetesRequestAuth = {
+          token: krc[c.name],
+        };
+        const cAuth = await this.configureAuth(c, authMetadata);
+        return cAuth;
+      }),
+    );
+
+    return clusterDetailsAuth;
+  }
+
+  protected getClusterIP(details: ClusterDetails): string {
+    const client = this.getKubeConfig(details);
+    return client.getCurrentCluster()?.server || '';
+  }
+
+  protected async makeRequestToCluster(
+    details: ClusterDetails,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const serverIP = this.getClusterIP(details);
+    if (!serverIP) {
+      this.logger.error(
+        `could not find server IP for cluster with name ${details.name}`,
+      );
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: null,
+      };
+    }
+
+    const query = decodeURIComponent(req.params.encodedQuery) || '';
+    const uri = `${serverIP}/${query}`;
+
+    const res = await this.sendClusterRequest(
+      details,
+      uri,
+      req.method,
+      req.body,
+    );
+
+    return res;
+  }
+
+  protected async sendClusterRequest(
+    details: ClusterDetails,
+    uri: string,
+    method: string,
+    body?: any,
+  ): Promise<KubernetesProxyResponse> {
+    const bearerToken = details.serviceAccountToken;
+    if (!bearerToken) {
+      this.logger.error('could not find bearer token!');
+      return {
+        code: ERROR_BAD_REQUEST,
+        data: {
+          error: 'Invalid service account token',
+        },
+      };
+    }
+
+    const ca = details.caData;
+    if (!ca) {
+      this.logger.error('could not find certificate!');
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: {
+          error: 'Invalid CA certificate configured within Backstage',
+        },
+      };
+    }
+
+    const agent = new https.Agent({ ca });
```

Comment:
the `https.Agent` constructor's parameters wind up getting [passed](https://github.com/nodejs/node/blob/a3e110820ff98702e1761831e7beaf0f5f1f75e7/lib/https.js#L1500) to `tls.connect`, which ultimately [passes](https://github.com/nodejs/node/blob/9cde7a033e66a8e9e691c708cfb448161fa2c27d/lib/internal/tls/secure-context.js#L62) the `ca` parameter to the C++ function `SecureContext::AddCACert`, where it gets [parsed](https://github.com/nodejs/node/blob/5fad0b93667ffc6e4def52996b9529ac99b26319/src/crypto/crypto_context.cc#L702-L703) by the OpenSSL function `PEM_read_bio_X509_AUX`, which according to [the docs](https://www.openssl.org/docs/man1.1.1/man3/PEM_read_bio_X509_AUX.html)
> read(s) ... structures in PEM format. In this sense PEM format is simply base64 encoded data surrounded by header lines

However, when backstage [sets up a KubeConfig](https://github.com/backstage/backstage/blob/a370706a18a3b8e3db8822b231a0bb554c8faf6c/plugins/kubernetes-backend/src/service/KubernetesClientProvider.ts#L49-L54) object from `@kubernetes/client-node` normally, that object will [set HTTPS options](https://github.com/kubernetes-client/javascript/blob/6131baeb1ca46337597b458a0b73ccc0261cfd9b/src/config.ts#L406-L409) on k8s API requests by first [base64-decoding](https://github.com/kubernetes-client/javascript/blob/6131baeb1ca46337597b458a0b73ccc0261cfd9b/src/config.ts#L513-L515) the value from `caData`.

All that to say, can you make sure to base64-decode the `caData` value here rather than using it verbatim?

Reply:
Also I'll be the first to admit that the [docs on this topic](https://backstage.io/docs/features/kubernetes/configuration#clusterscadata-optional), which say
> PEM-encoded certificate authority certificates
feel a bit misleading for not mentioning the extra level of base64-encoding that is expected (even though that is subtly implied by the "inspecting the Kubernetes config file" instruction)

Reply:
I took a [stab](https://github.com/backstage/backstage/pull/12911) at fixing those docs